pkg/webhook/bd_replacement.go

```diff
@@ -0,0 +1,383 @@
+// is already present in CSPC.
+func (bdr *BlockDeviceReplacement) IsNewBDPresentOnCurrentCSPC(newRG apis.RaidGroup, oldRG apis.RaidGroup) bool {
+	newBDs := GetNewBDFromRaidGroups(newRG, oldRG)
+	for _, pool := range bdr.OldCSPC.Spec.Pools {
+		for _, rg := range pool.RaidGroups {
+			for _, bd := range rg.BlockDevices {
+				if _, ok := newBDs[bd.BlockDeviceName]; ok {
+					return true
+				}
+			}
+		}
+	}
+	return false
+}
+
+// IsExistingReplacmentInProgress returns true if a block device in raid group is under active replacement.
+func (bdr *BlockDeviceReplacement) IsExistingReplacmentInProgress(oldRG apis.RaidGroup) (bool, error) {
+	for _, v := range oldRG.BlockDevices {
+		err, bdcObject := bdr.GetBDCOfBD(v.BlockDeviceName)
+		if err != nil {
+			return true, errors.Errorf("failed to query for any existing replacement in the raid group : %s", err.Error())
+		}
+		if bdcObject.HasAnnotationKey(apis.PredecessorBDKey) {
+			return true, errors.Errorf("replacement is still in progress for bd %s", v.BlockDeviceName)
+		}
+	}
+	return false, nil
+}
+
+// AreNewBDsValid returns true if the new BDs are valid BDs for replacement.
+func (bdr *BlockDeviceReplacement) AreNewBDsValid(newRG apis.RaidGroup, oldRG apis.RaidGroup, oldcspc *apis.CStorPoolCluster) bool {
+	newBDs := GetNewBDFromRaidGroups(newRG, oldRG)
+	for new, _ := range newBDs {
+		err, bdc := bdr.GetBDCOfBD(new)
+		if err != nil {
+			return false
+		}
+		if !bdr.IsBDValid(new, bdc, oldcspc) {
+			return false
+		}
+	}
+	return true
+}
+
+// IsBDValid returns true if the new BD is a valid BD for replacement.
+func (bdr *BlockDeviceReplacement) IsBDValid(bd string, bdc *bdc.BlockDeviceClaim, oldcspc *apis.CStorPoolCluster) bool {
+	if bdc != nil && !bdc.HasLabel(string(apis.CStorPoolClusterCPK), oldcspc.Name) {
+		return false
+	}
+	err, predecessorMap := bdr.GetPredecessorBDIfAny(oldcspc)
+	if err != nil {
+		return false
+	}
+	if predecessorMap[bd] {
+		return false
+	}
+	return true
+}
+
+// GetPredecessorBDIfAny returns a map of predecessor BDs if any in the current CSPC
+// Note: Predecessor BDs in a CSPC are those BD for which a new BD has appeared in the CSPC and
+//       replacement is still in progress
+//
+// For example,
+// (b1,b2) is a group in cspc
+// which has been changed to ( b3,b2 )  [Notice that b1 got replaced by b3],
+// now b1 is not present in CSPC but the replacement is still in progress in background.
+// In this case b1 is a predecessor BD.
+func (bdr *BlockDeviceReplacement) GetPredecessorBDIfAny(cspcOld *apis.CStorPoolCluster) (error, map[string]bool) {
+	predecessorBDMap := make(map[string]bool)
+	for _, pool := range cspcOld.Spec.Pools {
+		for _, rg := range pool.RaidGroups {
+			for _, bd := range rg.BlockDevices {
+				err, bdc := bdr.GetBDCOfBD(bd.BlockDeviceName)
+				if err != nil {
+					return err, nil
+				}
+				predecessorBDMap[bdc.Object.GetAnnotations()[apis.PredecessorBDKey]] = true
```

Comment:
It will crash when `bdc` is nil(a case where BDC is not yet created)

Reply:
done